sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -1502,4 +1502,24 @@ class AdaptiveQueryExecSuite
       }
     }
   }
+
+  test("SPARK-38030: Query with cast containing non-nullable columns should succeed with AQE") {
```

Comment:
Is there a reason why you provide this at `branch-3.1` PR only?

Reply:
We encountered this issue in production when running Spark 3.1.1. However, this repro does not work on 3.2+. This is because the code which would previously trigger access on an unresolved object is [now lazy](https://github.com/apache/spark/blob/7e5c3b216431b6a5e9a0786bf7cded694228cdee/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala#L132) and hence does not trigger the issue in 3.2+, although the root cause still existed in 3.2+. So I only added this test in 3.1.

Reply:
Could you put that explanation into the PR description?

Reply:
Added an explanation in the "How was this patch tested?" section.